File: src/main/java/org/veupathdb/lib/blast/field/FormatType.java
package org.veupathdb.lib.blast.field;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import com.fasterxml.jackson.databind.JsonNode;
import org.veupathdb.lib.blast.util.DefaultingJSONValue;
import org.veupathdb.lib.blast.util.JSONConstructor;

public enum FormatType implements DefaultingJSONValue
{
  Pairwise,
  QueryAnchoredShowingIdentities,
  QueryAnchoredNoIdentities,
  FlatQueryAnchoredShowingIdentities,
  FlatQueryAnchoredNoIdentities,
  BlastXML,
  Tabular,
  TabularWithCommentLines,
  SeqAlignTextASN1,
  SeqAlignBinaryASN1,
  CommaSeparatedValues,
  BlastArchiveASN1,
  SeqAlignJSON,
  MultipleFileBlastJSON,
  MultipleFileBlastXML2,
  SingleFileBlastJSON,
  SingleFileBlastXML2,
  SequenceAlignmentMap,
  OrganismReport,
  ;

  public int getValue() {
    return this.ordinal();
  }

  @Override
  public boolean isDefault() {
    return this == Pairwise;
  }

  @Override
  public String toString() {
    return String.valueOf(getValue());
  }

  @Override
  @JsonValue
  public JsonNode toJSON() {
    return JSONConstructor.newInt(this.ordinal());
  }

  public static FormatType fromIntValue(int value) {
    return FormatType.values()[value];
  }

  @JsonCreator
  public static FormatType fromJSON(JsonNode js) {
    return fromIntValue(js.intValue());
  }
}

File: src/main/java/org/veupathdb/lib/blast/field/FormatType.kt
package org.veupathdb.lib.blast.field

import org.veupathdb.lib.blast.util.DefaultingJSONValue
import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.JsonNode
import org.veupathdb.lib.blast.util.JSONConstructor
import com.fasterxml.jackson.annotation.JsonCreator

enum class FormatType : DefaultingJSONValue {
  Pairwise,
  QueryAnchoredShowingIdentities,
  QueryAnchoredNoIdentities,
  FlatQueryAnchoredShowingIdentities,
  FlatQueryAnchoredNoIdentities,
  BlastXML,
  Tabular,
  TabularWithCommentLines,
  SeqAlignTextASN1,
  SeqAlignBinaryASN1,
  CommaSeparatedValues,
  BlastArchiveASN1,
  SeqAlignJSON,
  MultipleFileBlastJSON,
  MultipleFileBlastXML2,
  SingleFileBlastJSON,
  SingleFileBlastXML2,
  SequenceAlignmentMap,
  OrganismReport;

  val value = ordinal

  override val isDefault
    get() = this == Pairwise

  override fun toString() = value.toString()

  @JsonValue
  override fun toJSON(): JsonNode = JSONConstructor.newInt(ordinal)

  companion object {
    @JvmStatic
    fun fromIntValue(value: Int): FormatType {
      return values()[value]
    }

    @JvmStatic
    @JsonCreator
    fun fromJSON(js: JsonNode): FormatType {
      return fromIntValue(js.intValue())
    }
  }
}